src/Compilers/Core/Portable/InternalUtilities/RoslynIncrementalHash.cs

```diff
@@ -0,0 +1,169 @@
+
+        if (_resetPending)
+        {
+            _hash.Initialize();
+            _resetPending = false;
+        }
+
+        _hash.TransformBlock(data, offset, count, null, 0);
+    }
+
+    /// <summary>Retrieves the hash for the data accumulated from prior calls to the <see cref="AppendData(byte[])"/>
+    /// method, and resets the object to its initial state.</summary>
+    /// <returns>The computed hash or HMAC.</returns>
+    /// <exception cref="ObjectDisposedException">The <see cref="RoslynIncrementalHash"/> object has already been disposed.</exception>
+    public byte[] GetHashAndReset()
+    {
+        if (_disposed)
+        {
+            throw new ObjectDisposedException(typeof(RoslynIncrementalHash).Name);
+        }
+
+        if (_resetPending)
+        {
+            _hash.Initialize();
+        }
+
+        _hash.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
+        byte[] hash = _hash.Hash!;
+        _resetPending = true;
+        return hash;
+    }
+
+    /// <summary>Releases the resources used by the current instance of the <see cref="RoslynIncrementalHash"/> class.</summary>
+    public void Dispose()
+    {
+        _disposed = true;
+        if (_hash != null)
+        {
+            _hash.Dispose();
+            _hash = null!;
+        }
+    }
+
+    /// <summary>Creates a <see cref="RoslynIncrementalHash"/> for the specified algorithm.</summary>
+    /// <param name="hashAlgorithm">The name of the hash algorithm to perform.</param>
+    /// <returns>An <see cref="RoslynIncrementalHash"/> instance ready to compute the hash algorithm specified by <paramref name="hashAlgorithm"/>.</returns>
+    /// <exception cref="ArgumentException"><paramref name="hashAlgorithm"/>.<see cref="HashAlgorithmName.Name"/> is <see langword="null"/> or an empty string.</exception>
+    /// <exception cref="CryptographicException"><paramref name="hashAlgorithm" /> is not a known hash algorithm.</exception>
+    public static RoslynIncrementalHash CreateHash(HashAlgorithmName hashAlgorithm)
+    {
+        if (string.IsNullOrEmpty(hashAlgorithm.Name))
+        {
+            throw new ArgumentException("Cryptography_HashAlgorithmNameNullOrEmpty", nameof(hashAlgorithm));
+        }
+        return new RoslynIncrementalHash(hashAlgorithm, GetHashAlgorithm(hashAlgorithm));
+    }
+
+    private static HashAlgorithm GetHashAlgorithm(HashAlgorithmName hashAlgorithm)
+    {
+        if (hashAlgorithm == HashAlgorithmName.MD5)
+        {
+            return MD5.Create();
+        }
+
+        if (hashAlgorithm == HashAlgorithmName.SHA1)
+        {
+            return SHA1.Create();
+        }
+
+        if (hashAlgorithm == HashAlgorithmName.SHA256)
+        {
+            return SHA256.Create();
+        }
+
+        if (hashAlgorithm == HashAlgorithmName.SHA384)
+        {
+            return SHA384.Create();
+        }
+
+        if (hashAlgorithm == HashAlgorithmName.SHA512)
+        {
+            return SHA512.Create();
+        }
```

Comment:
📝 The change relative to the reference implementation is here. .NET Framework creates instances of `SHA256CryptoServiceProvider` instead of calling `SHA256.Create()` (likewise for each of the others).

Reply:
💭 In theory, we could also address this by modifying the way we create our pool of `IncrementalHash` objects to use reflection to reset the `_hash` field.

Reply:
and it's hte provider that was causing the problem before?